tools/clang_tidy/lib/clang_tidy.dart

```diff
@@ -171,23 +189,91 @@ class ClangTidy {
-      if (lintAction != LintAction.skipThirdParty && command.containsAny(changedFiles)) {
-        buildCommands.add(command);
+  /// Returns f(n) = value(n * [shardCount] + [id]).
+  Iterable<T> _takeShard<T>(Iterable<T> values, int id, int shardCount) sync* {
+    int count = 0;
+    for (final T val in values) {
+      if (count % shardCount == id) {
+        yield val;
+      }
+      count++;
+    }
+  }
+
+  /// This returns a `_SetStatusCommand` for each [Command] in [items].
+  /// `Intersection` if the Command shows up in [items] and its filePath in all
+  /// [filePathSets], otherwise `Difference`.
+  Iterable<_SetStatusCommand> _calcIntersection(
+      Iterable<Command> items, Iterable<Set<String>> filePathSets) sync* {
+    bool allSetsContain(Command command) {
+      for (final Set<String> filePathSet in filePathSets) {
+        if (!filePathSet.contains(command.filePath)) {
+          return false;
+        }
+      }
+      return true;
+    }
+    for (final Command command in items) {
+      if (allSetsContain(command)) {
+        yield _SetStatusCommand(_SetStatus.Intersection, command);
+      } else {
+        yield _SetStatusCommand(_SetStatus.Difference, command);
       }
     }
-    return buildCommands;
+  }
+
+  /// Given a build commands json file's contents in [buildCommandsData], and
+  /// the [files] with local changes, compute the lint commands to run.  If
+  /// build commands are supplied in [sharedBuildCommandsData] the intersection
+  /// of those build commands will be calculated and distributed across
+  /// instances via the [shardId].
+  @visibleForTesting
+  Future<List<Command>> getLintCommandsForFiles(
+    List<Object?> buildCommandsData,
+    List<io.File> files,
+    List<List<Object?>> sharedBuildCommandsData,
+    int? shardId,
+  ) {
+    final List<Command> totalCommands = <Command>[];
+    if (sharedBuildCommandsData.isNotEmpty) {
+      final Iterable<Command> buildCommands = buildCommandsData
+          .map((dynamic data) => Command.fromMap(data as Map<String, Object?>));
+      final Iterable<Set<String>> shardFilePaths =
+          sharedBuildCommandsData.map((List<Object?> list) => list
+              .map((Object? data) =>
+                  Command.fromMap((data as Map<String, Object?>?)!).filePath)
+              .toSet());
+      final Iterable<_SetStatusCommand> intersectionResults =
+          _calcIntersection(buildCommands, shardFilePaths);
+      totalCommands.addAll(intersectionResults
+          .where((_SetStatusCommand element) =>
+              element.setStatus == _SetStatus.Difference)
+          .map((_SetStatusCommand e) => e.command));
+      final List<Command> intersection = intersectionResults
+          .where((_SetStatusCommand element) =>
+              element.setStatus == _SetStatus.Intersection)
+          .map((_SetStatusCommand e) => e.command)
+          .toList();
+      // Make sure to sort results, not sure if there is a defined order in the json file.
```

Comment:
It matters because it's how the sharding scheme works.  We have to guarantee that all shards are starting with the same order to make sure that their jobs don't overlap.

Reply:
Okay, I think I'm missing something big. Could you explain a bit more about why the order matters, or what would go wrong if the jobs are out of order?

Reply:
Yep, so imagine we are sharding across iOS and mac.  They both generated their compile_commands.json files.  They look like this:
iOS
```
[
  { "path": "foo.cc" },
  { "path": "bar.cc" },
]
```
mac
```
[
  { "path": "bar.cc" },
  { "path": "foo.cc" },
]
```

If you ran mac as shard 0 and ios and shard 1 (where job(n) = n * 2 + shard), they will both just execute on "foo.cc".  We don't have a guarantee about the order of compile_command.json as far as I know, so they have to be sorted.

Reply:
Got it. This explanation should probably go in a comment in the code.

Reply:
expanded on the existing comment